Update graphql-upload types

felipelima-circle opened the pull request:
- `@types/graphql-upload` doesn't work with new version of the library
- Generate updated types from JSDoc types defined in the library
---
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/66678#issuecomment-1716093027
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

andrewbranch commented on types/graphql-upload/processRequest.d.mts:
```diff
@@ -1,19 +1,52 @@
+    options?: ProcessRequestOptions,
+): Promise<
+    Array<
+        {
+            [key: string]: unknown;
+        } | {
+            [key: string]: unknown;
+        }
```
Same type twice in a union?

felipelima-circle replied:
My bad while fixing the style. This should be `{ [key: string]: unknown } | Array<{ [key: string]: unknown }>`. Fixed.

andrewbranch commented on types/graphql-upload/processRequest.d.mts:
```diff
@@ -1,19 +1,52 @@
+    options?: ProcessRequestOptions | undefined,
+) => Promise<
+    Array<
+        {
+            [key: string]: unknown;
+        } | {
+            [key: string]: unknown;
+        }
```
Same here

andrewbranch reviewed (changes requested):
> doesn't work with new version of the library

Can you be more specific about what doesn’t work?

> Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/66678#issuecomment-1716093027

This is my bad I think. arethetypeswrong shows this package is clean except for the notice that it’s an ESM-only package, which shouldn’t count as a problem for DT.

So I’m not quite sure what exactly this PR is attempting to fix, but it looks like there are a number of stylistic changes that are making it a bit confusing to review. Additionally, there are some renames like `UploadOptions` → `ProcessRequestOptions` that make this a breaking change in a patch release. For a popular package, breaking changes are expected to be really important correctness fixes, not just tweaks and renames. I’m guessing this comes from the auto-generation process you mentioned, but it’s something we need to be careful of here.

felipelima-circle commented:
> Can you be more specific about what doesn’t work?

- The `FileUpload` interface is missing the `capacitor` property
- The `graphqlUploadExpress()` and the `graphqlUploadKoa()` functions are missing the new `processRequest` parameter

> it looks like there are a number of stylistic changes that are making it a bit confusing to review

This was caused by the auto-generation process, which follows the same definition order as the main repo. I thought it would be nice to keep it like that to make it easier for the next updates, but since we always need to make some small adjustments, I changed it to be similar to what we have now.

> Additionally, there are some renames like `UploadOptions` → `ProcessRequestOptions` that make this a breaking change in a patch release

Some types were renamed or moved around by the author. Not sure if we should have done the same on the previous major release, and we missed that. Anyway, I decided to keep them for now to avoid breaking changes, and added a comment to remove them on the next major release.

andrewbranch reviewed (approved):
Thanks!

felipelima-circle commented:
Ready to merge

Shinigami92 commented:
Seems like DefinitelyTyped has published that as 16.0.6, where it at least should have been a 16.1.0 because the types are now breaking.

I get errors like:

```ts
Type 'Promise<{ filename: string; mimetype: string; encoding: string; createReadStream: () => Readable; }>' is not assignable to type 'Promise<FileUpload>'.
  Property 'capacitor' is missing in type '{ filename: string; mimetype: string; encoding: string; createReadStream: () => internal.Readable; }' but required in type 'FileUpload'.

116     upload.promise = Promise.resolve({
        ~~~~~~~~~~~~~~

  node_modules/.pnpm/@types+graphql-upload@16.0.6/node_modules/@types/graphql-upload/processRequest.d.mts:41:5
    41     capacitor: WriteStream;
           ~~~~~~~~~
    'capacitor' is declared here.
```

I used `~16.0.5`, updated my lock file and now getting these errors.
For now I will pin to `16.0.5` so I can fix the type-issues later. At runtime everything is working fine for months.

felipelima-circle commented:
@Shinigami92 Sorry about that. It seems that `capacitor` is an internal property and can be removed from the types. I fixed that [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68085). Thanks for reporting.
